### ISSUE TITLE
redirect={false} does not work properly for Edit forms used in datagrid expand

### ISSUE THREAD

**Issue by TCM5** [marmelab/react-admin#4865]
**What you were expecting:**
As reported in documentation, the `redirect` prop defaults to List, but when used other options like `"edit"` or `={false}` the redirect should be reflected

**What happened instead:**
When used options like `"edit"` or `={false}`, they are not considered and the redirect allways goes to list view.

**Steps to reproduce:**

1. Create a new Edit view with a SimpleForm like this:

```
<Edit {...rest} aside={<Aside {...rest} />} undoable={false}>
  <SimpleForm
     className={classes.form}
     basePath={controllerProps.basePath}
     record={controllerProps.record}
     save={controllerProps.save}
     version={controllerProps.version}
     redirect={false}
     resource="users"> 

      ... omitted for brevity

  </SimpleForm>
</Edit>
```

2.  Submit the form. It will not stay in the edit view, it will redirect to List view

**Environment**

* React-admin version:  3.4.0
* React version: 16.9.0
* Browser: Firefox 76.0.1 (64-bit)

**Comment by djhi** [comment]
Thanks for reporting this. If you are able to illustrate the bug or feature request with an example, please provide a sample application via one of the following means:

* CodeSandbox (https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple)
* A sample application via GitHub

**Comment by TCM5** [comment]
@djhi here is [https://codesandbox.io/s/gracious-cloud-189ym?file=/src/tags/TagEdit.js](https://codesandbox.io/s/gracious-cloud-189ym?file=/src/tags/TagEdit.js)

I changed the redirect prop on TagEdit from "list" to "edit", and the reported behaviour occurs.

**Comment by fzaninotto** [comment]
`redirect="edit"` does not make sens on an Edit view. As seen in the docs:

> Possible values are “edit”, “show”, “list”, and false to disable redirection

You have to use `redirect={false}` in this case - and it works.

**Comment by djhi** [comment]
I confirm `show` and `{false}` work on your codesandbox.

**Comment by dnvarun** [comment]
I am facing the same issue. please take a look at this sandbox: https://codesandbox.io/s/nameless-worker-dhryj?file=/src/tags/TagEdit.js
I have made the same change as OP. Set `redirect={false}` in TagEdit.js. But still the redirection occurs.

**Comment by dnvarun** [comment]
I am facing the same issue. I have deleted my previous comment as my earlier sandbox suddenly started working when I opened the webpage after the sandbox was dormant for a while. 

I have created a new sandbox which closely mirrors my layout, and the error is reproducible now. Please take a look at https://codesandbox.io/s/nice-dawn-95btl?file=/src/posts/PostEdit.js
I have put the PostEdit view in the expand on rowClick. 

When I go to the second page of Posts and do any edit and save, it redirects to the first page of Posts **after a few seconds**, even though I have set `redirect={false}`

**Comment by fzaninotto** [comment]
Reproduced, tanks for the sandbox. 

It seems this was introduced by #4422. In your particular case (an Edit view used as Expand), the `redirect={false]` triggers a `refresh(false)`, and the `useRefresh` code wipes out the location search. We need to do a better job in `useRefresh` .

Note that, until this gets fixed, you can use the `<Edit onSubmit>` prop (see #4881) as soon as version 3.6 is released.

**Comment by fzaninotto** [comment]
In  fact, I think it's the same problem as #4902. The fix for that bug should solve this one.

**Comment by fzaninotto** [comment]
Fixed by #4936
